### PR TITLE
Making `DataGridView` icons more clear

### LINKED ISSUE #7722: [HC] The icon of "AddNew" in DataGridView control is unclear
### .NET version

.NET 8.0.100-alpha.1.22451.15

### Did it work in .NET Framework?

No

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

No, it is not a regression issue

### Issue description

**Current Behavior:**
The icon of "AddNew" in DataGridView control is unclear.
HC Aquatic theme:
data-Aquatic
HC Dusk theme:
data-Dusk
HC Night Sky theme:
data-Night sky
**Expected Behavior:**
The icon of "AddNew" in DataGridView control should be clear.
win11 HC Desert theme:
data-Desert
win10 HC White theme:
data-HC White

### Steps to reproduce

1. Turn on High Contrast theme in Windows settings
2. Create a WinForms .NET core project 
3. Add a DataGridView control
4. Expand smart tag and click "Add column..." to add a column
5. Ctrl+Shift+A add a Application Manifest File and uncomment windows 10
6. Build and run
7. Observe the icon of "AddNew"

**More info:**
1. This issue cannot repro on win11 HC Desert theme. 
2. This issue is repro on win10 machine but cannot repro on win10 HC White theme. 
3. This issue can repro on framework project.

### PR DESCRIPTION
**Pull request by NikitaSemenovAkvelon**
Fixes #7722

## Proposed changes

- Added a `ControlPaint.CreateBitmapWithInvertedForeColor` method for inverting colors of the bitmap if the image's primary color has low contrast with a background.
- Used the `ControlPaint.CreateBitmapWithInvertedForeColor` method for `DataGridViewRowHeaderCell` images if HC mode is on.
- Covered the `ControlPaint.CreateBitmapWithInvertedForeColor` method with unit tests.

## Customer Impact

- Images in `DataGridViewRowHeaderCell` become more clear.

## Regression? 

- No

## Risk

- Minimal

### Before

image
image
image
image

### After

Screenshot 2023-01-16 112352
Screenshot 2023-01-16 112342
Screenshot 2023-01-16 112256
Screenshot 2023-01-16 112245
Screenshot 2023-01-16 103801
Screenshot 2023-01-16 103750
Screenshot 2023-01-16 112136
Screenshot 2023-01-16 112122

## Test methodology 

- Manually
- AI
- Unit testing

## Accessibility testing  

- AI

 

## Test environment(s) 

- .NET SDK:
 Version:   8.0.100-alpha.1.22512.5
 Commit:    1b80461e45

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.22621

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/dotnet/winforms/pull/8507)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewRowHeaderCell.cs [hunk 1]
```diff
@@ -1103,6 +1103,11 @@ private void PaintIcon(Graphics g, Bitmap bmp, Rectangle bounds, Color foreColor
 
             ImageAttributes attr = new ImageAttributes();
             attr.SetRemapTable(s_colorMap, ColorAdjustType.Bitmap);
+            if (SystemInformation.HighContrast)
+            {
+                bmp = ControlPaint.CreateBitmapWithInvertedForeColor(bmp, Color.Black, backColor);
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewRowHeaderCell.cs:1108]: I don't like hard-coding `Color.Black` as `imageMainColor` because icons could be changed to have other colors in the future, and also this method accepts non-specific bitmaps and could be used to draw icons with different colors. But I don't see any good alternatives... Maybe you have some ideas?

File: src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs [hunk 2]
```diff
@@ -2479,6 +2479,36 @@ internal static void InvertForeColorIfNeeded(Bitmap bitmap, Color backgroundColo
             }
         }
 
+        internal static Bitmap CreateBitmapWithInvertedForeColor(Bitmap bitmap, Color imageMainColor, Color backgroundColor)
+        {
+            if (Math.Abs(backgroundColor.GetBrightness() - imageMainColor.GetBrightness()) > .5)
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs:2484]: This condition seems arbitrary, maybe we could base it on WCAG color contrast ratio guideline (3:1)?

File: src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs [hunk 3]
```diff
@@ -2479,6 +2479,36 @@ internal static void InvertForeColorIfNeeded(Bitmap bitmap, Color backgroundColo
...
+                    }
+
+                    if (!inverted)
+                    {
+                        result.SetPixel(x, y, pixel);
+                    }
+                }
+            }
+
+            return result;
```
[the remark this answers is not in the document]
Comment (NikitaSemenovAkvelon) [reply src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs:2509]: We shouldn't do that. At least we don't understand what is happening with the object behind this method. And what is more important in this case is, the original image is stored in memory and is used multiple times. Disposing would cause an exception.

File: src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs [hunk 4]
```diff
@@ -2479,6 +2479,36 @@ internal static void InvertForeColorIfNeeded(Bitmap bitmap, Color backgroundColo
...
+                    if (pixel != backgroundColor)
+                    {
+                        result.SetPixel(x, y, pixel.InvertColor());
+                        inverted = true;
+                    }
+
+                    if (!inverted)
+                    {
+                        result.SetPixel(x, y, pixel);
+                    }
```
[the remark this answers is not in the document]
Comment (NikitaSemenovAkvelon) [reply src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs:2505]: Oh, it has been left from the previous version of changes. Fixed it, thanks.

File: src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs (hunk 2 above)
Comment (NikitaSemenovAkvelon) [reply src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs:2484]: I've found a similar condition here:
https://github.com/dotnet/winforms/blob/f6eaa898b9f130b99b6249161e8133a6b91f8f8b/src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs#L2219
So, I just reused existed solution instead of creating another one.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewRowHeaderCell.cs (hunk 1 above)
Comment (NikitaSemenovAkvelon) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewRowHeaderCell.cs:1108]: No, it's the best that I've thought up. Every image sent to this method has similar colors. We can't replace black with white and vice versa as in other cases due to the colors of images are not exactly black and white. Also, we can't make a decision of inverting every pixel by comparing it with a background because it caused affects images.

Comment (dkazennov) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewRowHeaderCell.cs:1108]: > Maybe you have some ideas?

We can check for main color of image the same way we check if a pixel should be inverted in CreateBitmapWithInvertedForeColor. Iterate through pixels of the bitmap. Then ignore Cell background pixels and white contour (if the background is black) and get the color of rest of pixels. But this is probably not worth it?

Comment (NikitaSemenovAkvelon) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewRowHeaderCell.cs:1108]: > We can check for main color of image the same way we check if a pixel should be inverted in CreateBitmapWithInvertedForeColor. Iterate through pixels of the bitmap. Then ignore Cell background pixels and white contour (if the background is black) and get the color of rest of pixels. But this is probably not worth it?

I've considered this option. And no, we can't. I describe above why. First of all, images don't contain simple colors like white (255,255,255) or black (0,0,0). And luminosity used in CreateBitmapWithInvertedForeColor isn't the same as luminance which is calculated by WCAG guidelines (https://www.w3.org/TR/2008/REC-WCAG20-20081211/#relativeluminancedef).

Comment (dkazennov) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewRowHeaderCell.cs:1108]: > First of all, images don't contain simple colors like white (255,255,255) or black (0,0,0).

Ok, but we can use the color of pixels those are (1) not the cell background pixels and (2) most numerous among the rest of the pixels (pixels grouped by a color). Just an idea.

File: src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs (hunk 2 above)
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs:2484]: Consider refactoring this test into a static method in `ControlPaint` class to be reused elsewhere.

File: src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs [hunk 5]
```diff
@@ -2479,6 +2479,26 @@ internal static void InvertForeColorIfNeeded(Bitmap bitmap, Color backgroundColo
             }
         }
 
+        internal static Bitmap CreateBitmapWithInvertedForeColor(Bitmap bitmap, Color imageMainColor, Color backgroundColor)
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs:2482]: `Create` methods new-up a resource and return it for the caller to dispose. This is not the case with this method.  It might or might not had new-ed up the bitmap. Consider refactoring it.

File: src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs:2499]: It still contributes to the growth of handle count unnecessarily.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewRowHeaderCell.cs [hunk 6]
```diff
@@ -1103,6 +1103,11 @@ private void PaintIcon(Graphics g, Bitmap bmp, Rectangle bounds, Color foreColor
 
             ImageAttributes attr = new ImageAttributes();
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewRowHeaderCell.cs:1104]: Not part of you change, but if you have a chance, could you please realign code right above so that the ternary operator stands out? For example
```
int width = DataGridView.RightToLeftInternal
    ? bounds. Right - RowHeaderIconMarginWidth - s_iconsWidth
    : bounds.Left + RowHeaderIconMarginWidth;

    Rectangle bmpRect = new(width, bounds.Y + (bounds. Height - s_iconsHeight) / 2, s_iconsWidth, s_iconsHeight);
```

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewRowHeaderCell.cs (hunk 1 above)
Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewRowHeaderCell.cs:1108]: Please add a comment to this code that explains why the color is hardcoded.

Comment (Tanya-Solyanik) [reply src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewRowHeaderCell.cs:1108]: Please add parameter name as well.

**Review by Tanya-Solyanik — DISMISSED** [review]
Looks good, but had you tested it on windows10? If you don't have a machine, please send privates to @Olina-Zhang.

File: src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs [hunk 7]
```diff
@@ -396,6 +396,8 @@ private static DashStyle BorderStyleToDashStyle(ButtonBorderStyle borderStyle)
         /// </summary>
         internal static bool IsDarker(Color c1, Color c2) => new HLSColor(c1).Luminosity < new HLSColor(c2).Luminosity;
 
+        internal static bool IsBright(Color color) => color.GetBrightness() > .5;
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs:399]: Both call sites invoke this method with a `!`, consider reverting logic and calling it `IsDark` to get rid of negation in the `if` conditions.

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ControlPaintTests.cs [hunk 8]
```diff
@@ -264,6 +264,25 @@ public void ControlPaint_CreateHBitmapTransparencyMask_NullBitmap_ThrowsArgument
...
+        }
+
+        [WinFormsTheory]
+        [MemberData(nameof(CreateBitmapWithInvertedForeColor_TestData))]
+        public void ControlPaint_CreateBitmapWithInvertedForeColor(Color imageColor, Color expected)
+        {
+            using var bitmap = new Bitmap(1, 1);
+            bitmap.SetPixel(0, 0, imageColor);
+
+            using Bitmap newBitmap = ControlPaint.CreateBitmapWithInvertedForeColor(bitmap, Color.LightGray);
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ControlPaintTests.cs:280]: Consider: `newBitmap` -> `invertedBitmap`

**Review by Tanya-Solyanik — COMMENTED** [review]
Looks good, but let's wait for Olina's results and please see my nitpicks.

**Comment by Olina-Zhang** [comment]
Tested this PR on Win10 and Win11 with all High contrast modes, now the display of icons looks same as default theme, no new issue found.

File: src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewRowHeaderCell.cs [hunk 9]
```diff
@@ -1090,20 +1090,35 @@ protected override object GetValue(int rowIndex)
...
             s_colorMap[0].NewColor = foreColor;
             s_colorMap[0].OldColor = Color.Black;
 
             ImageAttributes attr = new ImageAttributes();
             attr.SetRemapTable(s_colorMap, ColorAdjustType.Bitmap);
-            g.DrawImage(bmp, bmpRect, 0, 0, s_iconsWidth, s_iconsHeight, GraphicsUnit.Pixel, attr);
+
+            if (SystemInformation.HighContrast &&
+                // We can't replace black with white and vice versa as in other cases due to the colors of images are not exactly black and white.
+                // Also, we can't make a decision of inverting every pixel by comparing it with a background because it caused affects images.
```
Comment (dmitrii-drobotov) [inline src/System.Windows.Forms/src/System/Windows/Forms/DataGridViewRowHeaderCell.cs:1109]: Can you rephrase the last part of this sentence? "because it caused affects images" is not very clear.

File: src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs [hunk 10]
```diff
@@ -391,11 +391,13 @@ private static DashStyle BorderStyleToDashStyle(ButtonBorderStyle borderStyle)
...
         public static Color DarkDark(Color baseColor) => new HLSColor(baseColor).Darker(1.0f);
 
+        internal static bool IsDark(Color color) => color.GetBrightness() <= .5;
+
         /// <summary>
         ///  Returns true if the luminosity of <paramref name="c1"/> is less than <paramref name="c2"/>.
         /// </summary>
         internal static bool IsDarker(Color c1, Color c2) => new HLSColor(c1).Luminosity < new HLSColor(c2).Luminosity;
-
+
```
Comment (lonitra) [inline src/System.Windows.Forms/src/System/Windows/Forms/ControlPaint.cs:400]: This whitespace is causing build to complain

**Review by lonitra — COMMENTED** [review]
Minor comments, looks great!

File: src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ControlPaintTests.cs [hunk 11]
```diff
@@ -2013,6 +2032,21 @@ public void ControlPaint_FillReversibleRectangle_Invoke_Success(Rectangle rectan
             ControlPaint.FillReversibleRectangle(rectangle, backColor);
         }
 
+        public static IEnumerable<object[]> IsBright_TestData()
```
Comment (Tanya-Solyanik) [inline src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/ControlPaintTests.cs:2035]: Please rename to match the new method name 😄

**Review by dmitrii-drobotov — APPROVED** [review]

**Review by dkazennov — APPROVED** [review]
Looks good!

**Review by dreddy-work — APPROVED** [review]

**Review by Tanya-Solyanik — APPROVED** [review]
Thank you for all the follow ups!
